File: ming/0915/binary search.java
class Solution {
    /**
     * @param nums: The integer array.
     * @param target: Target to find.
     * @return: The first position of target. Position starts from 0.
     */
    public int binarySearch(int[] nums, int target) {
        int start = 0;
        int end = nums.length;
        while (start + 1 < end){
            int mid = start - (start - end) / 2;
            if (nums[mid] >= target){
                end = mid;
            }else{
                start = mid;
            }
        }
        
        if (nums[start] == target){
            return start;
        }
        if (nums[end] == target){
            return end;
        }
        return -1;
    }
}

File: mu_wang/9_15/binary search.py
class Solution:
    # @param nums: The integer array
    # @param target: Target number to find
    # @return the first position of target in nums, position start from 0 
    def binarySearch(self, nums, target):
        # write your code here
        low = 0
        high = len(nums)-1
        minMid = high
        found = False
        while low<=high:
            mid = low + (high-low)/2
            if target < nums[mid]:
                high = mid-1
            elif target> nums[mid]:
                low = mid+1
            else:
                minMid = min(mid,minMid)
                high = mid-1
                found = True
        if found:
            return minMid
        else:
            return -1